### ISSUE TITLE
ps server内存分布不均

### ISSUE THREAD

**Issue by simon1024** [alibaba/x-deeplearning#85]
实现了一个简单的lr模型，在分布式环境（4个ps，11个worker）训练，发现内存在ps之间分布不均衡。
4个ps内存占用情况分别为：
ps0：2.9G， ps1: 73G， ps2: 73G，ps3:2.9G
看着像是scheduler将emb分成了两个partition。
我的问题：
1.  这个分布结果是预期内的么？
2.  对于这种分布问题在xdl里应该怎么分析呢？xdl不支持tensorboard，那么有没有别的工具或者手段可以看到所有变量的分布情况呢？
3.  对于内置的几种调度策略（Aneal，balance等）能否详细说明下。

相关参数：
init_model_size:100000000,  add_probability:0.9, batch_size:1024, Placementer 相关参数都没有配置。

样本数据：4亿条，每个样本不到200个int64特征

 
相关代码如下：
<pre>
def train(options):
    reader = init_reader("r1", options.input_paths, options.epochs,
                         options.batch_size,options.thread_num)
    batch = reader.read()
    emb = xdl.embedding('emb', batch['sparse'], xdl.TruncatedNormal(stddev=0.001),
                         1, options.init_model_size, vtype='hash',
                        feature_add_probability=options.add_probability)

    loss,auc,accuracy = model(emb, batch['label'])

    train_op = xdl.Ftrl(0.05, l1_regularization_strength=0.03,
                        l2_regularization_strength=0.01).optimize()

    global_step = get_global_step().value
    log_hook = xdl.LoggerHook([loss, auc, accuracy, global_step],
                              "train metrics, loss:{0}, auc:{1}, accuracy:{2}, g_step:{3}", 10)

    sess = xdl.TrainSession(hooks=[log_hook])
    local_step = 0
    while not sess.should_stop():
        ret = sess.run([train_op, global_step])
        if ret is None:
            logging.warning("sess.run return None, input data maybe outof range")
            break
        g_step = ret[1]
        if options.steps and g_step >= options.steps:
            logging.info("reach steps limit")
            break
        local_step = local_step + 1
        if local_step % 100 == 0:
            logging.info("training, local step:%s, global_step:%s"%(local_step, g_step) )
    logging.info("finish, local step:%s, global_step:%s"%(local_step, g_step) )

@xdl.tf_wrapper(is_training=True)
def model(emb, labels):
    with tf.variable_scope("train"):                                                                                                                            
        logits = tf.reshape(emb, (-1,))
        labels = tf.reshape(labels, (-1,))
        loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=labels, logits=logits))

        y_pred = tf.sigmoid(logits)
        _, auc = tf.metrics.auc(labels=labels, predictions=y_pred)

        threshold_predictions = tf.to_float(tf.greater_equal(y_pred, 0.1))
        _, accuracy = tf.metrics.accuracy(labels=labels, predictions=threshold_predictions)

    return loss, auc, accuracy
</pre>

**Comment by yiling-dc** [comment]
1. 这个结果是预期内的，具体分析见下面第3条；
2. worker和scheduler的日志里面有具体的variable分配结果，可以查看一下；
3. 在考虑参数分配的时候，有这样几个因素：内存，IO，CPU使用量，variable数量（一次请求的socket数量）等等。其中，内存是一个约束条件，因为内存只要能满足运行的配置需要，不会影响整体的训练速度，其他你可以认为是优化条件。内置的anneal和balance分配策略的目标都是在满足内存约束的前提下，尽量去提高整体的吞吐，只是anneal策略是对所有variable进行一个粗略的预估，balance策略是在预跑阶段收集了variable的信息之后做出了比较精确的分配方案。具体细节比较复杂，可以自己参考下代码。

**Comment by yiling-dc** [comment]
因此，有些ps内存占用很高，是因为上面分布了一些hash类型的变量，如果样本里的id数目不多，那么这些变量在IO方面的占比就会比较低。

**Comment by codescv** [comment]
这样对集群调度不会造成困难吗？我们在给ps分配节点的时候，一般会指定ps要request多少内存，如果这样的话就没法事先预估ps的内存了

**Comment by yiling-dc** [comment]
> 这样对集群调度不会造成困难吗？我们在给ps分配节点的时候，一般会指定ps要request多少内存，如果这样的话就没法事先预估ps的内存了

可以事先预估的，只要是所有ps的内存加起来能满足所有variable的所需要的内存总和，Placement策略都会找到一个分配方案，只是说在ps内存富余的情况下，内存不会作为均衡的指标，只会作为一个硬性约束。
